### ISSUE TITLE
[WAGON-19] Make file upload more robust

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-wagon#239]
**[Mark R. Diggory](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mdiggory)** opened **[WAGON-19](https://issues.apache.org/jira/browse/WAGON-19?redirect=false)** and commented

We would recommend that for upload of files to a repository that the following cases be handled to provide greater robustness.

1.) All uploads be to a "staging" area, this staging area could be the same directory or a temp directory and would upload the file with the file name extension of

Henk Penning comments:

> That would be great.
> 
> I think, the best way for adding/replace stuff is
> 
> - write a 'temp'
> - rename 'temp' to 'file'
> 
> because a rename is truly atomic if 'temp' and 'file' are
> in the same file system.
> 
> If you can implement the 'temp' for 'file' to be,
> for instance, '.tmp.file', I can easily teach the checkers
> to ignore '.tmp.*' files. I think rsync does something
> like that (even better .tmp.$$.file).

So the goals here are to verify that rsync handles ".tmp.$$.file" which will stop it from attempting to sync partial uploads. Henk can alter the md5 checking utilities at Apache to postpone checking .tmp files md5 signatures.

2.) All file permissions on uploaded files would best handled to be only writable by the individual user, not writable by group and readable by all. All directory permissions should be writable for user and group and readable by all. This forces the following implementation to be required.

Any file upload that attempts to overwrite a file should instead, move that file out of the way to a temporary location, upload to the new file using strategy (1) and then name it to the old file, once this is completed the old file can be removed. This provides a means be which file "ownership" can be determined and maintained. The problem this solves is the following, if files are "group writable" then any individual in the group can overwite the file altering its contents, historically we cannot tell who actually made the alteration. If there are concerns about the integrity of the artifact or its signature, it is unclear who was responsible for the alteration.

-Mark Diggory

---

**Attachments:**
- [wagon.patch](https://issues.apache.org/jira/secure/attachment/12720820/wagon.patch) (_9.18 kB_)
- [WAGON-19.patch](https://issues.apache.org/jira/secure/attachment/12720671/WAGON-19.patch) (_15.80 kB_)
- [WebDavWagon.patch](https://issues.apache.org/jira/secure/attachment/12720801/WebDavWagon.patch) (_1.40 kB_)

**Issue Links:**
- [MNG-1905](https://issues.apache.org/jira/browse/MNG-1905) Delete remote files before uploading them to avoid permission problems

- [MDEPLOY-28](https://issues.apache.org/jira/browse/MDEPLOY-28) Deployment should delete remote files and create new ones instead of modifying them
 (_**"is depended upon by"**_)

9 votes, 9 watchers

**Comment by jira-importer** [comment]
**[Henry S. Isidro](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hisidro)** commented

Here's a patch for wagon-ftp that uploads a file with '.tmp.$$.file' appended to it's filename. Once transfer is complete, it removes the temporary extension.

**Comment by jira-importer** [comment]
**[Henry S. Isidro](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hisidro)** commented

I deleted the wagon-ftp patch since I refactored it. This new patch includes implementations for strategy (1) for wagon-ftp, wagon-ssh, wagon-ssh-external and wagon-file.

**Comment by jira-importer** [comment]
**[Henry S. Isidro](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=hisidro)** commented

Here's one for wagon-webdav which is in the sandbox.

**Comment by jira-importer** [comment]
**[Carlos Sanchez Gonzalez](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlos)** commented

The part about permissions is covered by MDEPLOY-28

**Comment by jira-importer** [comment]
**[Carlos Sanchez Gonzalez](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlos)** commented

This an updated and improved patch. Requires a lot of testing before applying as this is a really core part of maven

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This issue has been auto closed because it has been inactive for a long period of time. If you think this issue still applies, retest your problem with the most recent version of Maven and the affected component, reopen and post your results.
